test/core/ping.spec.js

```diff
@@ -0,0 +1,212 @@
+      })
+    })
+
+    // Connect the nodes
+    before(function (done) {
+      this.timeout(60 * 1000)
+      ipfsdA.api.swarm.connect(bMultiaddr, done)
+    })
+
+    after((done) => ipfsdA.stop(done))
+    after((done) => ipfsdB.stop(done))
+
+    it('sends the specified number of packets', (done) => {
+      let packetNum = 0
+      const count = 3
+      pull(
+        ipfsdA.api.pingPullStream(ipfsdBId, { count }),
+        drain(({ Success, Time, Text }) => {
+          expect(Success).to.be.true()
+          // It's a pong
+          if (Time) {
+            packetNum++
+          }
+        }, (err) => {
+          expect(err).to.not.exist()
+          expect(packetNum).to.equal(count)
+          done()
+        })
+      )
+    })
+
+    it('pinging an unknown peer will fail accordingly', (done) => {
+      let messageNum = 0
+      const count = 2
+      pull(
+        ipfsdA.api.pingPullStream('unknown', { count }),
+        drain(({ Success, Time, Text }) => {
+          messageNum++
+          // Assert that the ping command falls back to the peerRouting
+          if (messageNum === 1) {
+            expect(Text).to.include('Looking up')
+          }
+
+          // Fails accordingly while trying to use peerRouting
+          if (messageNum === 2) {
+            expect(Success).to.be.false()
+          }
+        }, (err) => {
+          expect(err).to.not.exist()
+          expect(messageNum).to.equal(count)
+          done()
+        })
+      )
+    })
+  })
+
+  describe('DHT enabled', function () {
+    // Our bootstrap process will run 3 IPFS daemons where
+    // A ----> B ----> C
+    // Allowing us to test the ping command using the DHT peer routing
+    let ipfsdA
+    let ipfsdB
+    let ipfsdC
+    let bMultiaddr
+    let cMultiaddr
+    let ipfsdCId
+
+    // Spawn nodes
+    before(function (done) {
+      this.timeout(60 * 1000)
+
+      parallel([
+        spawnNode.bind(null, { dht: true }),
+        spawnNode.bind(null, { dht: true }),
+        spawnNode.bind(null, { dht: true })
+      ], (err, ipfsd) => {
+        expect(err).to.not.exist()
+        ipfsdA = ipfsd[0]
+        ipfsdB = ipfsd[1]
+        ipfsdC = ipfsd[2]
+        done()
+      })
+    })
+
+    // Get the peer info objects
+    before(function (done) {
+      this.timeout(60 * 1000)
+
+      parallel([
+        ipfsdB.api.id.bind(ipfsdB.api),
+        ipfsdC.api.id.bind(ipfsdC.api)
+      ], (err, peerInfo) => {
+        expect(err).to.not.exist()
+        bMultiaddr = peerInfo[0].addresses[0]
+        ipfsdCId = peerInfo[1].id
+        cMultiaddr = peerInfo[1].addresses[0]
+        done()
+      })
+    })
+
+    // Connect the nodes
+    before(function (done) {
+      this.timeout(60 * 1000)
+
+      parallel([
+        ipfsdA.api.swarm.connect.bind(ipfsdA.api, bMultiaddr),
+        ipfsdB.api.swarm.connect.bind(ipfsdB.api, cMultiaddr)
+      ], (err) => setTimeout(() => done(err), 500)) // FIXME timeout needed for connections to succeed  
```

Comment:
Instead of doing this, we should do a setInterval that times out after ~30 seconds or something like that. Inside the interval, we should check if both peers are connected with the other peer, and if they are, cancel the interval + timeout and move on. If the they are not connected after 30 seconds, fail the test. This makes the test rely less on artificial pauses in the test execution.

Reply:
@JGAntunes ^^

Reply:
Sure, I'll pick this up meanwhile, as soon as I have some time.

Reply:
Should be fixed now @VictorBjelkholm @diasdavid, let me know what you think - https://github.com/ipfs/js-ipfs/pull/1342/files#diff-0e68633477fac5c3c897f68c6b28f7a9R177